File: platform_compiler_erlang/src/flink_generator.erl
%%% @doc
%%% Flink SQL Code Generator.
%%% Generates Flink SQL code for Streaming mode.
%%% Supports Kubernetes deployment via Flink Operator.
%%% @end

-module(flink_generator).

-export([generate_source_ddl/1, generate_sink_ddl/1, generate_transformations/1,
         generate_udf_registrations/1, generate_k8s_manifest/1]).

-include("schemas.hrl").

%%====================================================================
%% API functions
%%====================================================================

%% @doc Generate Flink source table DDL.
-spec generate_source_ddl(map()) -> string().
generate_source_ddl(State) ->
    DSL = maps:get(dsl, State),
    Config = maps:get(config, State),
    Sources = maps:get(sources, maps:get(connectors, Config, #{}), []),
    
    Header = "-- ==========================================\n-- SOURCE TABLES\n-- ==========================================",
    TablesDDL = [generate_source_table(T, Sources, State) || T <- maps:get(tables, DSL, [])],
    string:join([Header | TablesDDL], "\n\n").

%% @doc Generate Flink sink table DDL.
-spec generate_sink_ddl(map()) -> string().
generate_sink_ddl(State) ->
    Config = maps:get(config, State),
    Sinks = maps:get(sinks, maps:get(connectors, Config, #{}), []),
    
    Header = "-- ==========================================\n-- SINK TABLES\n-- ==========================================",
    SinksDDL = [generate_sink_table(S, State) || S <- Sinks],
    string:join([Header | SinksDDL], "\n\n").

%% @doc Generate transformation logic (views and queries).
-spec generate_transformations(map()) -> string().
generate_transformations(State) ->
    DSL = maps:get(dsl, State),
    
    Header = "-- ==========================================\n-- TRANSFORMATIONS\n-- ==========================================",
    ViewsDDL = [generate_view(Name, Query, State) || {Name, Query} <- maps:to_list(maps:get(views, DSL, #{}))],
    InsertsDDL = [translate_to_flink(I) ++ ";" || I <- maps:get(inserts, DSL, [])],
    string:join([Header | ViewsDDL ++ InsertsDDL], "\n\n").

%% @doc Generate UDF registration code for Flink.
-spec generate_udf_registrations(map()) -> string().
generate_udf_registrations(State) ->
    DSL = maps:get(dsl, State),
    
    Header = "-- ==========================================\n-- UDF/UDTF REGISTRATIONS\n-- ==========================================",
    ModelsDDL = [generate_model_registration(M) || M <- maps:get(models, DSL, [])],
    string:join([Header | ModelsDDL], "\n\n").

%% @doc Generate Kubernetes manifest for Flink deployment.
-spec generate_k8s_manifest(map()) -> string().
generate_k8s_manifest(State) ->
    Config = maps:get(config, State),
    Meta = maps:get(meta, Config, #{}),
    K8s = maps:get(kubernetes, maps:get(orchestration, Config, #{}), undefined),
    Exec = maps:get(execution, Config, #{}),
    Memory = maps:get(memory, Exec, #{}),
    Cpu = maps:get(cpu, Exec, #{}),
    
    JobName = re:replace(maps:get(<<"name">>, Meta, <<"flink-pipeline">>), "-", "-", [global, {return, list}]),
    Namespace = case K8s of
        undefined -> "default";
        _ -> binary_to_list(maps:get(<<"namespace">>, K8s, <<"default">>))
    end,
    Image = case K8s of
        undefined -> "flink:1.17";
        _ -> binary_to_list(maps:get(<<"image">>, K8s, <<"flink:1.17">>))
    end,
    
    Parallelism = maps:get(parallelism, Exec, 1),
    JmMemory = case Memory of
        #{} -> binary_to_list(maps:get(<<"jobmanager">>, Memory, <<"2048m">>));
        _ -> "2048m"
    end,
    TmMemory = case Memory of
        #{} -> binary_to_list(maps:get(<<"taskmanager">>, Memory, <<"4096m">>));
        _ -> "4096m"
    end,
    CpuCores = case Cpu of
        #{} -> integer_to_list(maps:get(cores, Cpu, 2));
        _ -> "2"
    end,
    
    CheckpointInterval = case maps:get(checkpointing, Exec, undefined) of
        #{interval := Interval} -> binary_to_list(Interval);
        _ -> "60000"
    end,
    
    CheckpointPath = case maps:get(checkpointing, Exec, undefined) of
        #{path := Path} -> binary_to_list(Path);
        _ -> "file:///tmp/flink-checkpoints"
    end,
    
    io_lib:format(
"apiVersion: flink.apache.org/v1beta1
kind: FlinkDeployment
metadata:
  name: ~s
  namespace: ~s
spec:
  image: ~s
  flinkVersion: v1_17
  flinkConfiguration:
    taskmanager.numberOfTaskSlots: \"~B\"
    state.backend: rocksdb
    state.checkpoints.dir: \"~s\"
    execution.checkpointing.interval: \"~sms\"
  serviceAccount: ~s
  jobManager:
    resource:
      memory: ~s
      cpu: ~s
  taskManager:
    resource:
      memory: ~s
      cpu: ~s
  job:
    jarURI: local:///opt/flink/usrlib/~s.jar
    parallelism: ~B
    upgradeMode: stateless",
        [JobName, Namespace, Image, Parallelism, CheckpointPath, CheckpointInterval,
         case K8s of undefined -> "flink"; _ -> binary_to_list(maps:get(<<"service_account">>, K8s, <<"flink">>)) end,
         JmMemory, CpuCores, TmMemory, CpuCores, JobName, Parallelism]).

%%====================================================================
%% Internal functions
%%====================================================================

-spec generate_source_table(map(), [map()], map()) -> string().
generate_source_table(Table, Sources, State) ->
    Name = maps:get(name, Table),
    Connector = find_connector(Name, Sources),
    
    case Connector of
        undefined ->
            create_flink_table(Table);
        _ ->
            ConnType = binary_to_list(maps:get(<<"type">>, Connector, <<"kafka">>)),
            ConnConfig = maps:get(<<"config">>, Connector, #{}),
            case ConnType of
                "kafka" -> generate_kafka_source(Table, ConnConfig, State);
                "hdfs" -> generate_file_source(Table, ConnConfig, State);
                "s3" -> generate_file_source(Table, ConnConfig, State);
                _ -> "-- Unsupported source type: " ++ ConnType
            end
    end.

-spec generate_sink_table(map(), map()) -> string().
generate_sink_table(Sink, State) ->
    Name = binary_to_list(maps:get(<<"name">>, Sink, <<"">>)),
    ConnType = binary_to_list(maps:get(<<"type">>, Sink, <<"kafka">>)),
    ConnConfig = maps:get(<<"config">>, Sink, #{}),
    Semantics = binary_to_atom(binary_to_list(maps:get(<<"semantics">>, Sink, <<"at_least_once">>)), utf8),
    
    case ConnType of
        "kafka" -> generate_kafka_sink(Name, ConnConfig, Semantics);
        "hdfs" -> generate_file_sink(Name, ConnConfig);
        "s3" -> generate_file_sink(Name, ConnConfig);
        _ -> "-- Unsupported sink type: " ++ ConnType
    end.

-spec generate_kafka_source(map(), map(), map()) -> string().
generate_kafka_source(Table, Config, _State) ->
    Name = maps:get(name, Table),
    Columns = string:join([K ++ " " ++ base_generator:format_type(V) || {K, V} <- maps:to_list(maps:get(columns, Table, #{}))], ",\n  "),
    Servers = string:join(binary_to_list(maps:get(<<"bootstrap_servers">>, Config, [<<"localhost:9092">>])), ","),
    Topic = binary_to_list(maps:get(<<"topic">>, Config, list_to_binary(Name))),
    Format = string:to_upper(binary_to_list(maps:get(<<"format">>, Config, <<"json">>))),
    StartupMode = binary_to_list(maps:get(<<"scan_startup_mode">>, Config, <<"latest-offset">>)),
    
    WatermarkClause = case maps:get(stream_config, Table, undefined) of
        undefined -> "";
        SC ->
            TimeAttr = maps:get(time_attribute, SC, ""),
            WmInterval = maps:get(watermark, SC, "5 seconds"),
            ",\n  WATERMARK FOR " ++ TimeAttr ++ " AS " ++ TimeAttr ++ " - INTERVAL '" ++ WmInterval ++ "'"
    end,
    
    io_lib:format(
"-- Kafka Source: ~s
CREATE TABLE ~s (
  ~s~s
) WITH (
  'connector' = 'kafka',
  'properties.bootstrap.servers' = '~s',
  'topic' = '~s',
  'scan.startup.mode' = '~s',
  'format' = '~s'
);", [Name, Name, Columns, WatermarkClause, Servers, Topic, StartupMode, string:to_lower(Format)]).

-spec generate_file_source(map(), map(), map()) -> string().
generate_file_source(Table, Config, _State) ->
    Name = maps:get(name, Table),
    Columns = string:join([K ++ " " ++ base_generator:format_type(V) || {K, V} <- maps:to_list(maps:get(columns, Table, #{}))], ",\n  "),
    DefaultPath = list_to_binary("/data/" ++ atom_to_list(Name)),
    Path = binary_to_list(maps:get(<<"path">>, Config, DefaultPath)),
    Format = string:to_upper(binary_to_list(maps:get(<<"format">>, Config, <<"parquet">>))),
    
    io_lib:format(
"-- File Source: ~s
CREATE TABLE ~s (
  ~s
) WITH (
  'connector' = 'filesystem',
  'path' = '~s',
  'format' = '~s'
);", [Name, Name, Columns, Path, string:to_lower(Format)]).

-spec generate_kafka_sink(string(), map(), atom()) -> string().
generate_kafka_sink(Name, Config, Semantics) ->
    Servers = string:join(binary_to_list(maps:get(<<"bootstrap_servers">>, Config, [<<"localhost:9092">>])), ","),
    Topic = binary_to_list(maps:get(<<"topic">>, Config, list_to_binary(Name))),
    
    SemanticValue = case Semantics of
        exactly_once -> "exactly-once";
        _ -> "at-least-once"
    end,
    
    io_lib:format(
"-- Kafka Sink: ~s
-- Use INSERT INTO to write to this sink
-- Configuration:
--   bootstrap.servers: ~s
--   topic: ~s
--   semantic: ~s", [Name, Servers, Topic, SemanticValue]).

-spec generate_file_sink(string(), map()) -> string().
generate_file_sink(Name, Config) ->
    DefaultPath = list_to_binary("/data/" ++ Name),
    Path = binary_to_list(maps:get(<<"path">>, Config, DefaultPath)),
    Format = string:to_upper(binary_to_list(maps:get(<<"format">>, Config, <<"parquet">>))),
    PartitionBy = maps:get(<<"partition_by">>, Config, []),
    
    PartitionClause = case PartitionBy of
        [] -> "";
        _ -> ",\n  'partition' = '" ++ string:join([binary_to_list(P) || P <- PartitionBy], ",") ++ "'"
    end,
    
    io_lib:format(
"-- File Sink: ~s
-- Use INSERT INTO to write to this sink
-- Configuration:
--   path: ~s
--   format: ~s~s", [Name, Path, string:to_lower(Format), PartitionClause]).

-spec create_flink_table(map()) -> string().
create_flink_table(Table) ->
    Name = maps:get(name, Table),
    Columns = string:join([K ++ " " ++ base_generator:format_type(V) || {K, V} <- maps:to_list(maps:get(columns, Table, #{}))], ",\n  "),
    
    WatermarkClause = case maps:get(stream_config, Table, undefined) of
        undefined -> "";
        SC ->
            TimeAttr = maps:get(time_attribute, SC, ""),
            WmInterval = maps:get(watermark, SC, "5 seconds"),
            ",\n  WATERMARK FOR " ++ TimeAttr ++ " AS " ++ TimeAttr ++ " - INTERVAL '" ++ WmInterval ++ "'"
    end,
    
    io_lib:format(
"-- Table: ~s
CREATE TABLE ~s (
  ~s~s
);", [Name, Name, Columns, WatermarkClause]).

-spec generate_model_registration(map()) -> string().
generate_model_registration(Model) ->
    Name = maps:get(name, Model),
    Path = maps:get(path, Model),
    InputSchema = maps:get(input_schema, Model),
    OutputSchema = maps:get(output_schema, Model),
    
    io_lib:format(
"-- Model: ~s
-- Path: ~s
-- Input: ~s, Output: ~s

-- Create Python UDF for ~s
-- CREATE TEMPORARY FUNCTION ~s 
-- AS 'udfs.~s_udf' 
-- LANGUAGE PYTHON;", [Name, Path, InputSchema, OutputSchema, Name, Name, Name]).

-spec generate_view(string(), string(), map()) -> string().
generate_view(Name, Query, _State) ->
    TranslatedQuery = translate_to_flink(Query),
    io_lib:format("CREATE VIEW ~s AS\n~s;", [Name, TranslatedQuery]).

-spec translate_to_flink(string()) -> string().
translate_to_flink(Query) ->
    Translations = [
        {"TUMBLE_START(", "TUMBLE_START("},
        {"TUMBLE_END(", "TUMBLE_END("},
        {"TUMBLE(", "TUMBLE("},
        {"HOP(", "HOP("},
        {"SESSION(", "SESSION("},
        {"CROSS JOIN LATERAL TABLE(", "CROSS JOIN LATERAL TABLE("},
        {"APPLY_MODEL(", "apply_model("},
        {"ENCODE_TEXT(", "encode_text("},
        {"LLM_GENERATE(", "llm_generate("},
        {"VECTOR_SEARCH(", "vector_search("}
    ],
    lists:foldl(fun({From, To}, Acc) -> re:replace(Acc, From, To, [global, {return, list}]) end, Query, Translations).

-spec find_connector(string(), [map()]) -> map() | undefined.
find_connector(_Name, []) ->
    undefined;
find_connector(Name, [C | Rest]) ->
    BinName = maps:get(<<"name">>, C, <<>>),
    case BinName =:= list_to_binary(Name) of
        true -> C;
        false -> find_connector(Name, Rest)
    end.

File: platform_compiler/compilers/flink_generator.py
"""
Flink SQL Code Generator

Generates Flink SQL code for Streaming mode.
Supports Kubernetes deployment via Flink Operator.
"""

from typing import Dict, Any, List

from .base import BaseCodeGenerator
from ..models.schemas import DSLLayer, PlatformConfig


class FlinkGenerator(BaseCodeGenerator):
    """Code generator for Apache Flink (Streaming)."""

    def _get_platform_name(self) -> str:
        return "flink"

    def generate_source_ddl(self) -> str:
        """Generate Flink source table DDL."""
        lines = ["-- ==========================================",
                 "-- SOURCE TABLES",
                 "-- =========================================="]
        
        sources = self.config.connectors.get('sources', [])
        
        for table in self.dsl.tables:
            # Find matching source connector
            connector = None
            for src in sources:
                if src.get('name') == table.name:
                    connector = src
                    break
            
            if connector:
                conn_type = connector.get('type', 'kafka')
                conn_config = connector.get('config', {})
                
                if conn_type == 'kafka':
                    ddl = self._generate_kafka_source(table, conn_config)
                elif conn_type in ['hdfs', 's3']:
                    ddl = self._generate_file_source(table, conn_config)
                else:
                    ddl = f"-- Unsupported source type: {conn_type}"
                
                lines.append(ddl)
            else:
                lines.append(self._create_flink_table(table))
        
        return '\n\n'.join(lines)

    def generate_sink_ddl(self) -> str:
        """Generate Flink sink table DDL."""
        lines = ["-- ==========================================",
                 "-- SINK TABLES",
                 "-- =========================================="]
        
        sinks = self.config.connectors.get('sinks', [])
        
        for sink in sinks:
            sink_name = sink.get('name', '')
            conn_type = sink.get('type', 'kafka')
            conn_config = sink.get('config', {})
            semantics = sink.get('semantics', 'at_least_once')
            
            if conn_type == 'kafka':
                ddl = self._generate_kafka_sink(sink_name, conn_config, semantics)
            elif conn_type in ['hdfs', 's3']:
                ddl = self._generate_file_sink(sink_name, conn_config)
            else:
                ddl = f"-- Unsupported sink type: {conn_type}"
            
            lines.append(ddl)
        
        return '\n\n'.join(lines)

    def generate_transformations(self) -> str:
        """Generate transformation logic (views and queries)."""
        lines = ["-- ==========================================",
                 "-- TRANSFORMATIONS",
                 "-- =========================================="]
        
        # Generate views
        for view_name, query in self.dsl.views.items():
            translated_query = self._translate_to_flink(query)
            lines.append(f"CREATE VIEW {view_name} AS\n{translated_query};")
        
        # Generate insert statements
        for insert in self.dsl.inserts:
            translated = self._translate_to_flink(insert)
            lines.append(translated + ";")
        
        return '\n\n'.join(lines)

    def generate_udf_registrations(self) -> str:
        """Generate UDF registration code for Flink."""
        lines = ["-- ==========================================",
                 "-- UDF/UDTF REGISTRATIONS",
                 "-- =========================================="]
        
        for model in self.dsl.models:
            lines.append(f"-- Model: {model.name}")
            lines.append(f"-- Path: {model.path}")
            lines.append(f"-- Input: {model.input_schema}, Output: {model.output_schema}")
            
            # Flink Python UDF registration
            lines.append(f"""
-- Create Python UDF for {model.name}
-- CREATE TEMPORARY FUNCTION {model.name} 
-- AS 'udfs.{model.name}_udf' 
-- LANGUAGE PYTHON;
""")
        
        return '\n\n'.join(lines)

    def generate_k8s_manifest(self) -> str:
        """Generate Kubernetes manifest for Flink deployment."""
        meta = self.config.meta
        k8s = self.config.orchestration.kubernetes
        exec_cfg = self.config.execution
        
        job_name = meta.get('name', 'flink-pipeline').replace('-', '-')
        namespace = k8s.namespace if k8s else 'default'
        image = k8s.image if k8s else 'flink:1.17'
        
        parallelism = exec_cfg.parallelism
        jobmanager_memory = exec_cfg.memory.jobmanager if exec_cfg.memory else '2048m'
        taskmanager_memory = exec_cfg.memory.taskmanager if exec_cfg.memory else '4096m'
        
        checkpoint_path = ''
        if exec_cfg.checkpointing and exec_cfg.checkpointing.path:
            checkpoint_path = exec_cfg.checkpointing.path
        
        manifest = f'''apiVersion: flink.apache.org/v1beta1
kind: FlinkDeployment
metadata:
  name: {job_name}
  namespace: {namespace}
spec:
  image: {image}
  flinkVersion: v1_17
  flinkConfiguration:
    taskmanager.numberOfTaskSlots: "{parallelism}"
    state.backend: rocksdb
    state.checkpoints.dir: "{checkpoint_path or 'file:///tmp/flink-checkpoints'}"
    execution.checkpointing.interval: "{exec_cfg.checkpointing.interval if exec_cfg.checkpointing else '60000'}ms"
  serviceAccount: {k8s.service_account if k8s and k8s.service_account else 'flink'}
  jobManager:
    resource:
      memory: {jobmanager_memory}
      cpu: {exec_cfg.cpu.cores if exec_cfg.cpu else 1}
  taskManager:
    resource:
      memory: {taskmanager_memory}
      cpu: {exec_cfg.cpu.cores if exec_cfg.cpu else 2}
  job:
    jarURI: local:///opt/flink/usrlib/{job_name}.jar
    parallelism: {parallelism}
    upgradeMode: stateless
'''
        return manifest

    def _generate_kafka_source(self, table, config: Dict[str, Any]) -> str:
        """Generate Kafka source DDL for Flink SQL."""
        columns = ',\n  '.join(f'{k} {self._format_type(v)}' for k, v in table.columns.items())
        servers = ','.join(config.get('bootstrap_servers', ['localhost:9092']))
        topic = config.get('topic', table.name)
        format_type = config.get('format', 'json').upper()
        startup_mode = config.get('scan_startup_mode', 'latest-offset')
        
        watermark_clause = ''
        if table.stream_config and table.stream_config.time_attribute:
            wm_interval = table.stream_config.watermark or '5 seconds'
            time_attr = table.stream_config.time_attribute
            watermark_clause = f',\n  WATERMARK FOR {time_attr} AS {time_attr} - INTERVAL \'{wm_interval}\''
        
        return f'''-- Kafka Source: {table.name}
CREATE TABLE {table.name} (
  {columns}{watermark_clause}
) WITH (
  'connector' = 'kafka',
  'properties.bootstrap.servers' = '{servers}',
  'topic' = '{topic}',
  'scan.startup.mode' = '{startup_mode}',
  'format' = '{format_type.lower()}'
);'''

    def _generate_file_source(self, table, config: Dict[str, Any]) -> str:
        """Generate file-based source DDL for Flink."""
        columns = ',\n  '.join(f'{k} {self._format_type(v)}' for k, v in table.columns.items())
        path = config.get('path', f'/data/{table.name}')
        format_type = config.get('format', 'parquet').upper()
        
        return f'''-- File Source: {table.name}
CREATE TABLE {table.name} (
  {columns}
) WITH (
  'connector' = 'filesystem',
  'path' = '{path}',
  'format' = '{format_type.lower()}'
);'''

    def _generate_kafka_sink(self, name: str, config: Dict[str, Any], semantics: str) -> str:
        """Generate Kafka sink DDL for Flink SQL."""
        # For sinks, we typically use INSERT INTO, so define as a regular table
        # The actual sink is defined by the INSERT statement
        servers = ','.join(config.get('bootstrap_servers', ['localhost:9092']))
        topic = config.get('topic', name)
        
        semantic_value = 'exactly-once' if semantics == 'exactly_once' else 'at-least-once'
        
        return f'''-- Kafka Sink: {name}
-- Use INSERT INTO to write to this sink
-- Configuration:
--   bootstrap.servers: {servers}
--   topic: {topic}
--   semantic: {semantic_value}'''

    def _generate_file_sink(self, name: str, config: Dict[str, Any]) -> str:
        """Generate file-based sink DDL for Flink."""
        path = config.get('path', f'/data/{name}')
        format_type = config.get('format', 'parquet').upper()
        partition_by = config.get('partition_by', [])
        
        partition_clause = ''
        if partition_by:
            partition_clause = f",\n  'partition' = '{','.join(partition_by)}'"
        
        return f'''-- File Sink: {name}
-- Use INSERT INTO to write to this sink
-- Configuration:
--   path: {path}
--   format: {format_type.lower()}{partition_clause}'''

    def _create_flink_table(self, table) -> str:
        """Create Flink table from schema definition."""
        columns = ',\n  '.join(f'{k} {self._format_type(v)}' for k, v in table.columns.items())
        
        watermark_clause = ''
        if table.stream_config and table.stream_config.time_attribute:
            wm_interval = table.stream_config.watermark or '5 seconds'
            time_attr = table.stream_config.time_attribute
            watermark_clause = f',\n  WATERMARK FOR {time_attr} AS {time_attr} - INTERVAL \'{wm_interval}\''
        
        return f'''-- Table: {table.name}
CREATE TABLE {table.name} (
  {columns}{watermark_clause}
);'''

    def _translate_to_flink(self, query: str) -> str:
        """Translate DSL-specific syntax to Flink SQL."""
        # Replace DSL window functions with Flink equivalents
        translations = [
            ('TUMBLE_START(', 'TUMBLE_START('),
            ('TUMBLE_END(', 'TUMBLE_END('),
            ('TUMBLE(', 'TUMBLE('),
            ('HOP(', 'HOP('),
            ('SESSION(', 'SESSION('),
            ('CROSS JOIN LATERAL TABLE(', 'CROSS JOIN LATERAL TABLE('),
            ('APPLY_MODEL(', 'apply_model('),
            ('ENCODE_TEXT(', 'encode_text('),
            ('LLM_GENERATE(', 'llm_generate('),
            ('VECTOR_SEARCH(', 'vector_search('),
        ]
        
        result = query
        for dsl_fn, flink_fn in translations:
            result = result.replace(dsl_fn, flink_fn)
        
        return result
